File: stream-chat-android-docs/src/main/java/io/getstream/chat/docs/java/client/cms/Messages.java
package io.getstream.chat.docs.java.client.cms;

import android.content.Context;

import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.getstream.chat.android.client.ChatClient;
import io.getstream.chat.android.client.api.models.PinnedMessagesPagination;
import io.getstream.chat.android.client.api.models.QueryChannelRequest;
import io.getstream.chat.android.client.channel.ChannelClient;
import io.getstream.chat.android.client.utils.ProgressCallback;
import io.getstream.chat.android.models.Attachment;
import io.getstream.chat.android.models.FilterObject;
import io.getstream.chat.android.models.Filters;
import io.getstream.chat.android.models.Message;
import io.getstream.chat.android.models.Reaction;
import io.getstream.chat.android.models.User;
import io.getstream.chat.android.models.querysort.QuerySortByField;
import io.getstream.chat.docs.java.client.helpers.MyFileUploader;
import io.getstream.result.Error;

public class Messages {
    private ChatClient client;
    private ChannelClient channelClient;
    private Message message;
    private Message parentMessage;

    /**
     * @see <a href="https://getstream.io/chat/docs/send_message/?language=java">Messages Overview</a>
     */
    class MessagesOverview {
        public void sendAMessage() {
            ChannelClient channelClient = client.channel("messaging", "general");
            Message message = new Message.Builder()
                    .withText("Josh, I told them I was pesca-pescatarian. Which is one who eats solely fish who eat other fish.")
                    .build();

            channelClient.sendMessage(message).enqueue(result -> {
                if (result.isSuccess()) {
                    Message sentMessage = result.getOrNull();
                } else {
                    // Handle error
                }
            });
        }

        public void sendAComplexMessage() {
            // Create an image attachment
            Map<String, Object> customData = new HashMap<>();
            customData.put("myCustomField", 123);
            Attachment attachment = new Attachment.Builder()
                    .withType("image")
                    .withImageUrl("https://bit.ly/2K74TaG")
                    .withThumbUrl("https://bit.ly/2Uumxti")
                    .withExtraData(customData)
                    .build();

            // Create a message with the attachment and a user mention
            Message message = new Message.Builder()
                    .withText("@Josh I told them I was pesca-pescatarian. Which is one who eats solely fish who eat other fish.")
                    .withAttachments(Collections.singletonList(attachment))
                    .withMentionedUsersIds(Collections.singletonList("josh-id"))
                    .withExtraData(Collections.singletonMap("anotherCustomField", 234))
                    .build();

            // Send the message to the channel
            channelClient.sendMessage(message).enqueue(result -> { /* ... */ });
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_message/?language=java#get-a-message">Get A Message</a>
         */
        public void getAMessage() {
            channelClient.getMessage("message-id").enqueue(result -> {
                if (result.isSuccess()) {
                    Message message = result.getOrNull();
                } else {
                    // Handle error
                }
            });
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_message/?language=java#update-a-message">Update A Message</a>
         */
        public void updateAMessage() {
            // Update some field of the message
            Message updatedMessage = message.newBuilder()
                    .withText("my updated text")
                    .build();

            // Send the message to the channel
            channelClient.updateMessage(updatedMessage).enqueue(result -> {
                if (result.isSuccess()) {
                    Message remoteUpdatedMessage = result.getOrNull();
                } else {
                    // Handle error
                }
            });
        }

        public void partialUpdateAMessage() {
            // Update some field of the message
            Message updatedMessage = message.newBuilder()
                    .withText("my updated text")
                    .build();

            // Send the message to the channel
            channelClient.updateMessage(updatedMessage).enqueue(result -> {
                if (result.isSuccess()) {
                    Message remoteUpdatedMessage = result.getOrNull();
                } else {
                    // Handle error
                }
            });
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_message/?language=java#delete-a-message">Delete A Message</a>
         */
        public void deleteAMessage() {
            channelClient.deleteMessage("message-id", false).enqueue(result -> {
                if (result.isSuccess()) {
                    Message deletedMessage = result.getOrNull();
                } else {
                    // Handle error
                }
            });
        }
    }

    class MessageFormat {
        /**
         * @see <a href="https://getstream.io/chat/docs/android/message_format/?language=java">Open Graph Scrapper</a>
         */
        public void openGraphScraper() {
            Message message = new Message.Builder()
                    .withText("Check this bear out https://imgur.com/r/bears/4zmGbMN")
                    .build();

            channelClient.sendMessage(message).enqueue(result -> { /* ... */ });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/file_uploads/?language=java">File Uploads</a>
     */
    class FileUploads {
        File imageFile = new File("path");
        File anyOtherFile = new File("path");

        public void fileUploads() {
            ChannelClient channelClient = client.channel("messaging", "general");

            // Upload an image without detailed progress
            channelClient.sendImage(imageFile).enqueue(result -> {
                if (result.isSuccess()) {
                    // Successful upload, you can now attach this image
                    // to a message that you then send to a channel
                    String imageUrl = result.getOrNull().getFile();

                    Attachment attachment = new Attachment.Builder()
                            .withImageUrl(imageUrl)
                            .withType("image")
                            .build();

                    Message message = new Message();
                    message.getAttachments().add(attachment);

                    channelClient.sendMessage(message).enqueue(res -> { /* ... */ });
                } else {
                    // Handle error
                }
            });

            // Upload a file, monitoring for progress with a ProgressCallback
            channelClient.sendFile(anyOtherFile, new ProgressCallback() {
                @Override
                public void onSuccess(@NotNull String file) {
                    String fileUrl = file;
                }

                @Override
                public void onError(@NotNull Error error) {
                    // Handle error
                }

                @Override
                public void onProgress(long bytesUploaded, long totalBytes) {
                    // You can render the uploading progress here
                }
            }).enqueue(); // No callback passed to enqueue, as we'll get notified above anyway
        }

        public void usingYourOwnCdn(String apiKey, Context context) {
            // Set a custom FileUploader implementation when building your client
            ChatClient client = new ChatClient.Builder("{{ api_key }}", context)
                    .fileUploader(new MyFileUploader())
                    .build();
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/android/file_uploads/?language=kotlin#deleting-files-and-images">Deleting Files and Images</a>
     */
    void deleteFileOrImage() {
        ChannelClient channelClient = client.channel("messaging", "general");

        // Deletes the image
        channelClient.deleteImage("{{ url of uploaded image }}").enqueue();

        // Deletes the file
        channelClient.deleteFile("{{ url of uploaded file }}").enqueue();
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/send_reaction/?language=java">Reactions</a>
     */
    class Reactions {
        public void sendAReaction() {
            ChannelClient channelClient = client.channel("messaging", "general");

            // Add reaction 'like' with a custom field
            Map<String, Object> customFields = new HashMap<>();
            customFields.put("customField", 1);
            Reaction reaction = new Reaction.Builder()
                    .withMessageId("message-id")
                    .withType("like")
                    .withScore(1)
                    .withExtraData(customFields)
                    .build();

            boolean enforceUnique = false; // Don't remove other existing reactions
            channelClient.sendReaction(reaction, enforceUnique).enqueue(result -> {
                if (result.isSuccess()) {
                    Reaction sentReaction = result.getOrNull();
                } else {
                    // Handle error
                }
            });

            // Add reaction 'like' and replace all other reactions of this user by it
            enforceUnique = true;
            channelClient.sendReaction(reaction, enforceUnique).enqueue(result -> {
                if (result.isSuccess()) {
                    Reaction sentReaction = result.getOrNull();
                } else {
                    // Handle error
                }
            });
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_reaction/?language=java#removing-a-reaction">Removing A Reaction</a>
         */
        public void removeAReaction() {
            String reactionType = "like";
            channelClient.deleteReaction("message-id", reactionType).enqueue(result -> {
                if (result.isSuccess()) {
                    Message message = result.getOrNull();
                } else {
                    // Handle error
                }
            });
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_reaction/?language=java#paginating-reactions">Paginating Reactions</a>
         */
        public void paginatingReactions() {
            // Get the first 10 reactions
            int offset = 0;
            int limit = 10;
            channelClient.getReactions("message-id", offset, limit).enqueue(result -> {
                if (result.isSuccess()) {
                    List<Reaction> reactions = result.getOrNull();
                } else {
                    // Handle error
                }
            });

            // Get the second 10 reactions
            offset = 10;
            channelClient.getReactions("message-id", offset, limit)
                    .enqueue(result -> { /* ... */ });
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_reaction/?language=java#cumulative-clap-reactions">Cumulative (Clap) Reactions</a>
         */
        public void cumulativeReactions() {
            Reaction reaction = new Reaction.Builder()
                    .withMessageId("message-id")
                    .withType("like")
                    .withScore(5)
                    .build();

            boolean enforceUnique = false;
            channelClient.sendReaction(reaction, enforceUnique).enqueue(result -> { /* ... */ });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/threads/?language=java">Threads & Replies</a>
     */
    class ThreadsAndReplies {
        public void startAThread() {
            Message message = new Message.Builder()
                    .withText("Hello there!")
                    .withParentId(parentMessage.getId())
                    .build();

            // Send the message to the channel
            channelClient.sendMessage(message).enqueue(result -> {
                if (result.isSuccess()) {
                    Message sentMessage = result.getOrNull();
                } else {
                    // Handle error
                }
            });
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/threads/?language=java#thread-pagination">Thread Pagination</a>
         */
        public void threadPagination() {
            int limit = 20;
            // Retrieve the first 20 messages inside the thread
            client.getReplies(parentMessage.getId(), limit).enqueue(result -> {
                if (result.isSuccess()) {
                    List<Message> replies = result.getOrNull();
                } else {
                    // Handle error
                }
            });

            // Retrieve the 20 more messages before the message with id "42"
            client.getRepliesMore(parentMessage.getId(), "42", limit).enqueue(result -> { /* ... */ });
        }

        Message originalMessage;

        public void quoteMessage() {
            Message message = new Message.Builder()
                    .withText("This message quotes another message!")
                    .withReplyMessageId(originalMessage.getId())
                    .build();

            channelClient.sendMessage(message).enqueue(result -> { /* ... */ });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/silent_messages/?language=java">Silent Messages</a>
     */
    class SilentMessages {
        User systemUser;

        public void silentMessage() {
            Message message = new Message.Builder()
                    .withText("You and Jane are now matched!")
                    .withUser(systemUser)
                    .withSilent(true)
                    .build();

            channelClient.sendMessage(message).enqueue(result -> { /* ... */ });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/search/?language=java">Search</a>
     */
    class Search {
        public void searchMessages() {
            int offset = 0;
            int limit = 10;

            FilterObject channelFilter = Filters.in("members", Collections.singletonList("john"));
            FilterObject messageFilter = Filters.autocomplete("text", "supercalifragilisticexpialidocious");

            client.searchMessages(
                    channelFilter,
                    messageFilter,
                    offset,
                    limit,
                    null,
                    null
            ).enqueue(result -> {
                if (result.isSuccess()) {
                    List<Message> messages = result.getOrNull().getMessages();
                } else {
                    // Handle error
                }
            });
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/search/?language=java">Pinned Messages</a>
     */
    class PinnedMessages {
        public void pinAndUnpinAMessage() {
            // Create pinned message
            Calendar calendar = Calendar.getInstance();
            calendar.set(2077, 1, 1);
            Date pinExpirationDate = calendar.getTime();

            Message message = new Message.Builder()
                    .withText("Hey punk")
                    .withPinned(true)
                    .withPinExpires(pinExpirationDate)
                    .build();

            channelClient.sendMessage(message).enqueue(result -> { /* ... */ });

            // Unpin message
            channelClient.unpinMessage(message).enqueue(result -> { /* ... */ });

            // Pin message for 120 seconds
            channelClient.pinMessage(message, 120).enqueue(result -> { /* ... */ });

            // Change message expiration to 2077
            channelClient.pinMessage(message, pinExpirationDate).enqueue(result -> { /* ... */ });

            // Remove expiration date from pinned message
            channelClient.pinMessage(message, null).enqueue(result -> { /* ... */ });
        }

        public void retrievePinnedMessages() {
            channelClient.query(new QueryChannelRequest()).enqueue(result -> {
                if (result.isSuccess()) {
                    List<Message> pinnedMessages = result.getOrNull().getPinnedMessages();
                } else {
                    // Handle error
                }
            });
        }

        public void paginateOverAllPinnedMessages() {
            // List the first page of pinned messages, pinned before now, of the channel with descending direction (newest on top)
            channelClient.getPinnedMessages(10, QuerySortByField.descByName("pinnedAt"), new PinnedMessagesPagination.BeforeDate(new Date(), false))
                    .enqueue(result -> {
                        if (result.isSuccess()) {
                            List<Message> pinnedMessages = result.getOrNull();
                        } else {
                            // Handle error
                        }
                    });

            // You can use a pinnedAt date retrieved from the previous request to get the next page
            Date nextDate = new Date();
            // List the next page of pinned messages
            channelClient.getPinnedMessages(10, QuerySortByField.descByName("pinnedAt"), new PinnedMessagesPagination.BeforeDate(nextDate, false))
                    .enqueue(result -> {
                        if (result.isSuccess()) {
                            List<Message> pinnedMessages = result.getOrNull();
                        } else {
                            // Handle error
                        }
                    });
        }
    }

    public class Translation {

        /**
         * @see <a href="https://getstream.io/chat/docs/translation/?language=java#message-translation-endpoint">Message Translation</a>
         */
        public void messageTranslation() {
            // Translate message to French
            ChannelClient channelClient = client.channel("messaging", "general");

            Message message = new Message.Builder()
                    .withText("Hello, I would like to have more information about your product.")
                    .build();

            channelClient.sendMessage(message).enqueue(result -> {
                if (result.isSuccess()) {
                    String messageId = result.getOrNull().getId();
                    client.translate(messageId, "fr").enqueue(translationResult -> {
                        if (result.isSuccess()) {
                            Message translatedMessage = result.getOrNull();
                            String translation = translatedMessage.getI18n().get("fr_text");
                        }
                         else {
                             // Handle error
                        }
                    });
                } else {
                    // Handle error
                }
            });
        }
    }
}

File: stream-chat-android-docs/src/main/kotlin/io/getstream/chat/docs/kotlin/client/cms/Messages.kt
package io.getstream.chat.docs.kotlin.client.cms

import android.content.Context
import io.getstream.chat.android.client.ChatClient
import io.getstream.chat.android.client.api.models.PinnedMessagesPagination
import io.getstream.chat.android.client.api.models.QueryChannelRequest
import io.getstream.chat.android.client.channel.ChannelClient
import io.getstream.chat.android.client.utils.ProgressCallback
import io.getstream.chat.android.models.Attachment
import io.getstream.chat.android.models.Filters
import io.getstream.chat.android.models.Message
import io.getstream.chat.android.models.Reaction
import io.getstream.chat.android.models.User
import io.getstream.chat.android.models.querysort.QuerySortByField
import io.getstream.chat.docs.kotlin.client.helpers.MyFileUploader
import io.getstream.result.Result
import io.getstream.result.Error
import java.io.File
import java.util.Calendar
import java.util.Date

class Messages(
    val client: ChatClient,
    val channelClient: ChannelClient,
    val message: Message,
    val parentMessage: Message,
) {

    /**
     * @see <a href="https://getstream.io/chat/docs/send_message/?language=kotlin">Messages Overview</a>
     */
    inner class MessagesOverview {
        fun sendAMessage() {
            val channelClient = client.channel("messaging", "general")
            val message = Message(
                text = "Josh, I told them I was pesca-pescatarian. Which is one who eats solely fish who eat other fish."
            )

            channelClient.sendMessage(message).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val sentMessage: Message = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }

            // Send a message synchronously
            val messageResult = channelClient.sendMessage(message).execute()
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/react/send_message/?language=kotlin#complex-example">Complex Example</a>
         */
        fun sendAComplexMessage() {
            // Create an image attachment
            val attachment = Attachment(
                type = "image",
                imageUrl = "https://bit.ly/2K74TaG",
                thumbUrl = "https://bit.ly/2Uumxti",
                extraData = mutableMapOf("myCustomField" to 123),
            )

            // Create a message with the attachment and a user mention
            val message = Message(
                text = "@Josh I told them I was pesca-pescatarian. Which is one who eats solely fish who eat other fish.",
                attachments = mutableListOf(attachment),
                mentionedUsersIds = mutableListOf("josh-id"),
                extraData = mutableMapOf("anotherCustomField" to 234),
            )

            // Send the message to the channel
            channelClient.sendMessage(message).enqueue { /* ... */ }
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_message/?language=kotlin#get-a-message">Get A Message</a>
         */
        fun getAMessage() {
            channelClient.getMessage("message-id").enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val message: Message = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_message/?language=kotlin#update-a-message">Update A Message</a>
         */
        fun updateAMessage() {
            // Update some field of the message
            val updatedMessage = message.copy(text = "my updated text")

            // Send the message to the channel
            channelClient.updateMessage(updatedMessage).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val updatedMessage: Message = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_message/?language=kotlin#delete-a-message">Delete A Message</a>
         */
        fun deleteAMessage() {
            channelClient.deleteMessage("message-id").enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val deletedMessage: Message = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }
    }

    inner class MessageFormat {
        /**
         * @see <a href="https://getstream.io/chat/docs/android/message_format/?language=kotlin">Open Graph Scrapper</a>
         */
        fun openGraphScraper() {
            val message = Message(
                text = "Check this bear out https://imgur.com/r/bears/4zmGbMN"
            )
            channelClient.sendMessage(message).enqueue { /* ... */ }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/file_uploads/?language=kotlin">File Uploads</a>
     */
    inner class FileUploads {
        val imageFile = File("path")
        val anyOtherFile = File("path")

        fun fileUploads() {
            val channelClient = client.channel("messaging", "general")

            // Upload an image without detailed progress
            channelClient.sendImage(imageFile).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        // Successful upload, you can now attach this image
                        // to a message that you then send to a channel
                        val imageUrl = result.value.file
                        val attachment = Attachment(
                            type = "image",
                            imageUrl = imageUrl,
                        )
                        val message = Message(
                            attachments = mutableListOf(attachment),
                        )
                        channelClient.sendMessage(message).enqueue { /* ... */ }
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }

            // Upload a file, monitoring for progress with a ProgressCallback
            channelClient.sendFile(
                anyOtherFile,
                object : ProgressCallback {
                    override fun onSuccess(url: String?) {
                        val fileUrl = url
                    }

                    override fun onError(error: Error) {
                        // Handle error
                    }

                    override fun onProgress(bytesUploaded: Long, totalBytes: Long) {
                        // You can render the uploading progress here.
                    }
                }
            ).enqueue() // No callback passed to enqueue, as we'll get notified above anyway
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/android/file_uploads/?language=kotlin#deleting-files-and-images">Deleting Files and Images</a>
         */
        fun deleteImageOrFile() {
            val channelClient = client.channel("messaging", "general")

            // Deletes the image
            channelClient.deleteImage("{{ url of uploaded image }}").enqueue()

            // Deletes the file
            channelClient.deleteFile("{{ url of uploaded file }}").enqueue()
        }

        fun usingYourOwnCdn(apiKey: String, context: Context) {
            // Set a custom FileUploader implementation when building your client
            val client = ChatClient.Builder("{{ api_key }}", context)
                .fileUploader(MyFileUploader())
                .build()
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/send_reaction/?language=kotlin">Reactions</a>
     */
    inner class Reactions {
        fun sendAReaction() {
            val channelClient = client.channel("messaging", "general")

            // Add reaction 'like' with a custom field
            val reaction = Reaction(
                messageId = "message-id",
                type = "like",
                score = 1,
                extraData = mutableMapOf("customField" to 1),
            )
            channelClient.sendReaction(reaction).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val sentReaction: Reaction = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }

            // Add reaction 'like' and replace all other reactions of this user by it
            channelClient.sendReaction(reaction, enforceUnique = true).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val sentReaction: Reaction = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_reaction/?language=kotlin#removing-a-reaction">Removing A Reaction</a>
         */
        fun removeAReaction() {
            channelClient.deleteReaction(
                messageId = "message-id",
                reactionType = "like",
            ).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val message = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_reaction/?language=kotlin#paginating-reactions">Paginating Reactions</a>
         */
        fun paginatingReactions() {
            // Get the first 10 reactions
            channelClient.getReactions(
                messageId = "message-id",
                offset = 0,
                limit = 10,
            ).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val reactions = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }

            // Get the second 10 reactions
            channelClient.getReactions(
                messageId = "message-id",
                offset = 10,
                limit = 10,
            ).enqueue { /* ... */ }
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/send_reaction/?language=kotlin#cumulative-clap-reactions">Cumulative (Clap) Reactions</a>
         */
        fun cumulativeReactions() {
            val reaction = Reaction(messageId = "message-id", type = "clap", score = 5)
            channelClient.sendReaction(reaction).enqueue { /* ... */ }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/threads/?language=kotlin">Threads & Replies</a>
     */
    inner class ThreadsAndReplies {
        fun startAThread() {
            val message = Message(
                text = "Hello there!",
                parentId = parentMessage.id,
            )

            // Send the message to the channel
            channelClient.sendMessage(message).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val sentMessage = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }

        /**
         * @see <a href="https://getstream.io/chat/docs/threads/?language=kotlin#thread-pagination">Thread Pagination</a>
         */
        fun threadPagination() {
            // Retrieve the first 20 messages inside the thread
            client.getReplies(parentMessage.id, limit = 20).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val replies = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }

            // Retrieve the 20 more messages before the message with id "42"
            client.getRepliesMore(
                messageId = parentMessage.id,
                firstId = "42",
                limit = 20,
            ).enqueue { /* ... */ }
        }

        val originalMessage = Message()

        fun quoteMessage() {
            val message = Message(
                text = "This message quotes another message!",
                replyMessageId = originalMessage.id,
            )
            channelClient.sendMessage(message).enqueue { /* ... */ }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/silent_messages/?language=kotlin">Silent Messages</a>
     */
    inner class SilentMessages {
        val systemUser = User()

        fun silentMessage() {
            val message = Message(
                text = "You and Jane are now matched!",
                user = systemUser,
                silent = true,
            )
            channelClient.sendMessage(message).enqueue { /* ... */ }
        }
    }

    /**
     * @see <a href="https://getstream.io/chat/docs/search/?language=kotlin">Search</a>
     */
    inner class Search {
        fun searchMessagesLimitAndOffsetPagination() {
            client.searchMessages(
                offset = 0,
                limit = 10,
                channelFilter = Filters.`in`("members", listOf("john")),
                messageFilter = Filters.autocomplete("text", "supercalifragilisticexpialidocious"),
            ).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val messages = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }
    }

    /**
     * @see <a https://getstream.io/chat/docs/pinned_messages/?language=java">Pinned Messages</a>
     */
    inner class PinnedMessages {

        fun pinAndUnpinAMessage() {
            // Create pinned message
            val pinExpirationDate = Calendar.getInstance().apply {
                set(2077, 1, 1)
            }.time
            val message = Message(
                text = "Hey punk",
                pinned = true,
                pinExpires = pinExpirationDate
            )

            channelClient.sendMessage(message).enqueue { /* ... */ }

            // Unpin message
            channelClient.unpinMessage(message).enqueue { /* ... */ }

            // Pin message for 120 seconds
            channelClient.pinMessage(message, timeout = 120).enqueue { /* ... */ }

            // Change message expiration to 2077
            channelClient.pinMessage(message, expirationDate = pinExpirationDate).enqueue { /* ... */ }

            // Remove expiration date from pinned message
            channelClient.pinMessage(message, expirationDate = null).enqueue { /* ... */ }
        }

        fun retrievePinnedMessages() {
            channelClient.query(QueryChannelRequest()).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val pinnedMessages = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }

        fun paginateOverAllPinnedMessages() {
            // List the first page of pinned messages, pinned before now, of the channel with descending direction (newest on top)
            channelClient.getPinnedMessages(
                limit = 10,
                sort = QuerySortByField.descByName("pinnedAt"),
                pagination = PinnedMessagesPagination.BeforeDate(
                    date = Date(),
                    inclusive = false,
                ),
            ).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val pinnedMessages = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }

            // You can use a pinnedAt date retrieved from the previous request to get the next page
            val nextDate = Date()
            // List the next page of pinned messages
            channelClient.getPinnedMessages(
                limit = 10,
                sort = QuerySortByField.descByName("pinnedAt"),
                pagination = PinnedMessagesPagination.BeforeDate(
                    date = nextDate,
                    inclusive = false,
                ),
            ).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val pinnedMessages = result.value
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }
    }

    inner class Translation {

        /**
         * @see <a href="https://getstream.io/chat/docs/translation/?language=kotlin#message-translation-endpoint">Message Translation</a>
         */
        fun messageTranslation() {
            // Translate message to French
            val channelClient = client.channel("messaging", "general")
            val message = Message(text = "Hello, I would like to have more information about your product.")

            channelClient.sendMessage(message).enqueue { result ->
                when (result) {
                    is Result.Success -> {
                        val messageId = result.value.id
                        val frenchLanguage = "fr"

                        client.translate(messageId, frenchLanguage).enqueue { translationResult ->
                            when (translationResult) {
                                is Result.Success -> {
                                    val translatedMessage = translationResult.value
                                    val translation = translatedMessage.getTranslation(frenchLanguage)
                                }
                                is Result.Failure -> {
                                    // Handle translationResult error
                                }
                            }
                        }
                    }
                    is Result.Failure -> {
                        // Handler error
                    }
                }
            }
        }
    }
}
